README.md
# AWS Lambda Router ![Build Status](https://github.com/Clorofila/aws-lambda-router/actions/workflows/publish.yml/badge.svg)

[![npm version](https://badge.fury.io/js/@clorofila%2Faws-lambda-router.svg)](https://badge.fury.io/js/@clorofila%2Faws-lambda-router)

A small library for [AWS Lambda](https://aws.amazon.com/lambda/details) providing routing for [API Gateway](https://aws.amazon.com/api-gateway), [Proxy Integrations](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-set-up-simple-proxy.html), [SNS](https://aws.amazon.com/sns) and [S3 Events](https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html).

## Features

-   Easy Handling of [ANY method](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-method-settings-method-request.html#setup-method-add-http-method) in API Gateways
-   Simplifies writing lambda handlers (in nodejs > 8)
-   Lambda Proxy Resource support for AWS API Gateway
-   Enable CORS for requests
-   No external dependencies - well, almost, only types of [aws-lambda](https://www.npmjs.com/package/@types/aws-lambda) :-)
-   Currently there are four `processors` (callers for Lambda) implemented:
    -   API Gateway ANY method (called proxyIntegration)
    -   SNS
    -   SQS
    -   S3
-   Compatibility with Typescript >= 3.5

## Installation

Install via npm

```
$ npm install aws-lambda-router
```

or yarn

```
$ yarn add aws-lambda-router
```

## Getting Started

This is a simple example of `aws-lambda-router` in conjunction with ANY method and the API Gateway proxy integration. The following code will respond with a message when executed using an AWS API Gateway with a `GET` request on URL path `<base-url-of-gateway>/gateway-mapping/article/123`. The corresponding AWS API Gateway Resource is `/article/{articleId}`.

```js
import * as router from 'aws-lambda-router'

export const handler = router.handler({
    proxyIntegration: {
        // assumes the first path part is `stage` and removes it.
        removeBasePath: true
        routes: [
            {
                // request-path-pattern with a path variable:
                path: '/article/:id',
                method: 'GET',
                // we can use the path param 'id' in the action call:
                action: (request, context) => {
                    return "You called me with: " + request.paths.id;
                }
            },
            {
                // request-path-pattern with a path variable in Open API style:
                path: '/section/{id}',
                method: 'GET',
                // we can use the path param 'id' in the action call:
                action: (request, context) => {
                    return "You called me with: " + request.paths.id;
                }
            }
        ]
    }
})
```

## Proxy path support (work in progress)

The proxy integration usually works using a path configured in the API gateway. For example: `/article/{id}`.

If you use the WIP _proxy path support_, the complete path will be used to match a route config in `proxyIntegration`. This can be used to build an [Simple Proxy with API Gateway](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-create-api-as-simple-proxy-for-http.html)

Example:

-   Resource in API Gateway : /{proxy+} [see Proxy Integration with a Proxy Resource](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-set-up-simple-proxy.html)
-   Method: ANY

With the lambda configuration shown below the following paths are matched:

-   _api-gateway-host_/article/list
-   _api-gateway-host_/api/json/v1/schema

```js
const router = require('aws-lambda-router');

exports.handler = router.handler({
    proxyIntegration: {
        proxyPath: proxy,
        routes: [
            {
                path: '/article/list',
                method: 'GET',
                action: (request, context) => {
                    return 'You called me with: ' + request.path;
                },
            },
            {
                path: '/api/json/v1/schema',
                method: 'GET',
                action: (request, context) => {
                    return 'You called me with: ' + request.path;
                },
            },
        ],
    },
});
```

Typescript example:

```ts
import * as router from 'aws-lambda-router'
import { ProxyIntegrationEvent } from 'aws-lambda-router/lib/proxyIntegration'

export const handler = router.handler({
    proxyIntegration: {
        removeBasePath: true,
        routes: [
            {
                path: '/saveExample',
                method: 'POST',
                // request.body needs type assertion, because it defaults to type unknown (user input should be checked):
                action: (request, context) => {
                    const { text } = request.body as { text: string }
                    return `You called me with: ${text}`
                }
            },
            {
                path: '/saveExample2',
                method: 'POST',
                // it's also possible to set a type (no type check):
                action: (request: ProxyIntegrationEvent<{ text: string }>, context) => {
                    return `You called me with: ${request.body.text}`
                }
            }
        ]
    }
}
```

## When an API Mapping is configured for a custom domain

If your API-gateway custom domain has a mapping for `stage` you will need to set `removeBasePath` to `false` in order to allow the route to be properly matched on.

```js
const router = require('aws-lambda-router');

exports.handler = router.handler({
    proxyIntegration: {
        removeBasePath: false,
        routes: [
            {
                path: '/api/article/list',
                method: 'GET',
                action: (request, context) => {
                    return 'You called me with: ' + request.path;
                },
            },
            {
                path: '/api/json/v1/schema',
                method: 'GET',
                action: (request, context) => {
                    return 'You called me with: ' + request.path;
                },
            },
        ],
    },
});
```

## Enable CORS

To activate CORS on all http methods (OPTIONS requests are handled automatically) you only need to set the parameter `cors` to `true` on the `proxyIntegration` rule.

See the following example:

<details>
  <summary>Default CORS example</summary>

```js
import * as router from 'aws-lambda-router';

export const handler = router.handler({
    // for handling an http-call from an AWS Apigateway proxyIntegration we provide the following config:
    proxyIntegration: {
        cors: true,
        routes: [
            {
                path: '/graphql',
                method: 'POST',
                // provide a function to be called with the appropriate data
                action: (request, context) => doAnything(request.body),
            },
        ],
    },
});
```

</details>

If CORS is activated, these default headers will be sent on every response:

    "Access-Control-Allow-Origin" = "'*'"
    "Access-Control-Allow-Methods" = "'GET,POST,PUT,DELETE,HEAD,PATCH'"
    "Access-Control-Allow-Headers" = "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'"

### Customizing CORS

To customize CORS for all routes pass any of the following options to the `proxyIntegration` `cors` property. If a property is not set then it will default to the above default CORS headers.

-   `origin`: Configures the **Access-Control-Allow-Origin** CORS header. Possible values:
    -   `Boolean` - set `origin` to `true` to reflect the request origin or set it to `false` to disable CORS.
    -   `String` - set `origin` to a specific origin. For example if you set it to `"http://example.com"` only requests from "http://example.com" will be allowed.
    -   `RegExp` - set `origin` to a regular expression pattern which will be used to test the request origin. If it's a match, the request origin will be reflected. For example the pattern `/example\.com$/` will reflect any request that is coming from an origin ending with "example.com".
    -   `Array` - set `origin` to an array of valid origins. Each origin can be a `String` or a `RegExp`. For example `["http://example1.com", /\.example2\.com$/]` will accept any request from "http://example1.com" or from a subdomain of "example2.com".
    -   `Function` - set `origin` to a function to be evaluated. The function will get passed the `APIGatewayProxyEvent` and must return the allowed origin or `false`
-   `methods`: Configures the **Access-Control-Allow-Methods** CORS header. Expects a comma-delimited string (ex: 'GET,PUT,POST') or an array (ex: `['GET', 'PUT', 'POST']`).
-   `allowedHeaders`: Configures the **Access-Control-Allow-Headers** CORS header. Expects a comma-delimited string (ex: 'Content-Type,Authorization') or an array (ex: `['Content-Type', 'Authorization']`). If not specified, defaults to reflecting the headers specified in the request's **Access-Control-Request-Headers** header.
-   `exposedHeaders`: Configures the **Access-Control-Expose-Headers** CORS header. Expects a comma-delimited string (ex: 'Content-Range,X-Content-Range') or an array (ex: `['Content-Range', 'X-Content-Range']`). If not specified, no custom headers are exposed.
-   `credentials`: Configures the **Access-Control-Allow-Credentials** CORS header. Set to `true` to pass the header, otherwise it is omitted.
-   `maxAge`: Configures the **Access-Control-Max-Age** CORS header. Set to an integer to pass the header, otherwise it is omitted.

<details>
  <summary>Customize CORS example</summary>

```js
import * as router from 'aws-lambda-router';

export const handler = router.handler({
    // for handling an http-call from an AWS Apigateway proxyIntegration we provide the following config:
    proxyIntegration: {
        cors: {
            origin: 'https://test.example.com', // Only allow CORS request from this url
            methods: ['GET', 'POST', 'PUT'], // Only allow these HTTP methods to make requests
        },
        routes: [
            {
                path: '/graphql',
                method: 'POST',
                // provide a function to be called with the appropriate data
                action: (request, context) => doAnything(request.body),
            },
        ],
    },
});
```

</details>

## Error mapping / handling

```js
import * as router from 'aws-lambda-router';

export const handler = router.handler({
    // for handling an http-call from an AWS Apigateway proxyIntegration we provide the following config:
    proxyIntegration: {
        routes: [
            {
                path: '/graphql',
                method: 'POST',
                action: (request, context) => doThrowAnException(request.body),
            },
        ],
        debug: true,
        errorMapping: {
            NotFound: 404,
            MyCustomError: 429,
            ServerError: 500,
        },
    },
});

function doThrowAnException(body) {
    throw { reason: 'MyCustomError', message: 'Throw an error for this example' };
}
```

With the key word `errorMapping` shown in the example above you can adjust the assignment of thrown errors to http response code error.
The action can throw an object like

    "throw {reason: 'NotFound', message: 'object id not found'}"

and the http response then contains the configured value as response code and the message as the body.

### Genereric error handler for proxyIntegration

Also there is implemented an generic error handler. The idea is to have a place for handling error logging and also return custom error messages.

<details>
  <summary>Generic error handling example</summary>

```js
onError: (error, event, context) => {
    // Global exceptions goes here, works for sns, s3 and sqs should end up here aswell
    console.log(error)
},
proxyIntegration: {
    onError: (error, request, context) => {
        // proxy integration exceptions goes here
        console.log('Error', error, request, context)
    },
    routes: ...
}

// Also support returning a response:
proxyIntegration: {
    onError: async (error) => {
        console.log('Error', error)
        await someAsyncMethod();
        return {
           statusCode: 401,
           body: Not allowed
        }
    },
    routes: ...
}
```

</details>

For more examples please look into the [tests](test/proxyIntegration.test.ts) of proxyIntegration.

## SNS to Lambda Integrations

SNS Event Structure: https://docs.aws.amazon.com/sns/latest/dg/sns-message-and-json-formats.html

For handling calls in Lambdas initiated from AWS-SNS you can use the following code snippet:

<details>
  <summary>SNS to Lambda example</summary>

```js
import * as router from 'aws-lambda-router';

export const handler = router.handler({
    sns: {
        routes: [
            {
                // a regex to match the content of the SNS-Subject:
                subject: /.*/,
                // Attention: the message is JSON-stringified
                action: (sns, context, records) => service.doSomething(JSON.parse(sns.Message)),
            },
        ],
    },
});
```

</details>

The _records_ parameter contains `SNSEventRecord[]`. An exampe event structure can be found [here](lib/event-examples/sns.json). For example you can parse now the [message attributes of the SNS](https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html) or reads the topic arn of SNS.

## SQS to Lambda Integrations

For handling calls in Lambdas initiated from AWS-SQS you can use the following code snippet:

<details>
  <summary>SQS to Lambda example</summary>

```js
import * as router from 'aws-lambda-router';

export const handler = router.handler({
    sqs: {
        routes: [
            {
                // match complete SQS ARN:
                source: 'arn:aws:sqs:us-west-2:594035263019:aticle-import',
                // Attention: the messages Array is JSON-stringified
                action: (messages, context, records) => messages.forEach(message => console.log(JSON.parse(message))),
            },
            {
                // a regex to match the source SQS ARN:
                source: /.*notification/,
                // Attention: the messages array is JSON-stringified
                action: (messages, context, records) => service.doNotify(messages),
            },
        ],
    },
});
```

</details>

An SQS message always contains an array of records. In each SQS record there is the message in the body JSON key.
The `action` method gets all body elements from the router as an array.

If more than one route matches, only the **first** is used!

The _records_ parameter contains the complete array of records, which handled by aws-lambda-router. An exampe can be found [here](lib/event-examples/sqs.json). This gives you the possibility to read metadata from the event. For example, you can parse the [message attributes of the SQS](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-message-attributes.html) and use them for further processing.

## S3 to Lambda Integrations

Lambdas can be triggered by S3 events. The router now supports these events.
With the router it is very easy and flexible to connect a lambda to different s3 sources (different buckets). The following configurations are available:

-   bucketName: By specifying a fixed _bucketName_ all s3 records with this bucket name are forwarded to a certain action. Instead of a fixed bucket a _RegExp_ is also possible.
-   eventName: By configuring the [S3 event name](https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html#supported-notification-event-types) the routing can be further restricted. A _RegExp_ is also possible here.
-   objectKeyPrefix: fixed string as an prefix of an object key (but not an RegExp). Is useful if you want to organize your bucket in subfolder.

A combination of bucketName, eventName and objectKeyPrefix is possible. If no _bucketName_, _eventName_ and _objectKeyPrefix_ is configured, all records of s3 events are forwarded to the action.

The action method will be called with the records of the [S3Event Structure](https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html)

The following examples demonstrates the most use cases:

<details>
  <summary>S3 to Lambda example</summary>

```js
import * as router from 'aws-lambda-router';

export const handler = router.handler({
    s3: {
        routes: [
            {
                //match every s3 record to this action
                action: (record, context) => console.log(record.s3.object.key, record.eventName),
            },
            {
                //match s3 events which created, bucket name is whitelisted here
                eventName: 'ObjectCreated:Put',
                action: (record, context) => console.log(record.s3.object.key, record.eventName),
            },
            {
                //event name is an regex: match 'ObjectCreated:Put' or 'ObjectCreated:Copy'
                eventName: /ObjectCreated:*/,
                action: (record, context) => console.log(record.s3.object.key, record.eventName),
            },
            {
                //exact name of bucket 'myBucket', event name is whitelisted and will not be checked
                bucketName: 'myBucket',
                action: (record, context) => console.log(record.s3.object.key, record.eventName),
            },
            {
                //regex of bucket name (all buckets started with 'bucket-dev-' will be machted
                bucketName: /^bucket-dev-.*/,
                action: (record, context) => console.log(record.s3.object.key, record.eventName),
            },
            {
                //action only will be called if bucket and event matched to the given regex
                bucketName: /bucket-dev-.*/,
                eventName: /ObjectCreated:*/,
                action: (record, context) => console.log(event.s3.object.key, record.eventName),
            },
            {
                //action only will be called if bucket and event matched to the given fixed string
                bucketName: 'bucket',
                eventName: 'ObjectRemoved:Delete',
                action: (record, context) => console.log(event.s3.object.key, record.eventName),
            },
            {
                //match if s3 events comes from Bucket 'bucket' with event 'ObjectRemoved:Delete'
                // and the object key starts with /upload
                objectKeyPrefix: '/upload',
                bucketName: 'bucket',
                eventName: 'ObjectRemoved:Delete',
                action: (record, context) => console.log(record.s3.object.key, record.eventName),
            },
        ],
        debug: true,
    },
});
```

</details>


CHANGELOG.md
### not released yet ()

### 0.13.4 (Aug 13, 2023)

-   Updated readme file with new badges.
-   Improved workflows descriptions.
-   Made proxyevent body non-optional

### 0.13.1 (Aug 11, 2023)

-   Publish package using Github actions

### 0.13.0 (Aug 11, 2023)

-   Prepare package for publishing on npm

### 0.12.2 (Aug 11, 2023)

-   Migrated from travis to github actions
-   Re-added clean script
-   Updated package codecov test dependency

### 0.12.1 (Aug 08, 2023)

-   Changed location of compiled files
-   Updated npm scripts
-   Updated package json to use new location of compiled files

### 0.12.0 (Aug 04, 2023)

-   Forked project
-   Added prettier support
-   Deprecated existing proxy integration
-   Added new proxy integration based on new aws gateway event (APIGatewayProxyEventV2)

### 0.11.0 (July 02, 2021)

-   improvement: add HttpMethod to "ProxyIntegrationRoute" ([#64](https://github.com/spring-media/aws-lambda-router/pull/64)) (thanks to [@JHPG](https://github.com/JHPG))
-   improvement: flag to disable Base Path Removal ([#72](https://github.com/spring-media/aws-lambda-router/pull/72)) (thanks to [@EthanDavis](https://github.com/EthanDavis))
-   fix: send cors headers back after onError result ([#69](https://github.com/spring-media/aws-lambda-router/pull/69)) (thanks to [@Gerjunior](https://github.com/Gerjunior))
-   dependency (dependabot): Bump hosted-git-info from 2.8.5 to 2.8.9 ([#68](https://github.com/spring-media/aws-lambda-router/pull/68))
-   dependency (dependabot): Bump lodash from 4.17.19 to 4.17.21 ([#67](https://github.com/spring-media/aws-lambda-router/pull/67))
-   dependency (dependabot): Bump minimist from 1.2.0 to 1.2.5 ([#73](https://github.com/spring-media/aws-lambda-router/pull/73))

### 0.10.0 (March 23, 2021)

-   adding generic and proxyIntegration errorHandler ([#60](https://github.com/spring-media/aws-lambda-router/pull/60)) (thanks to [@swaner](https://github.com/swaner))
-   fix: logging statements ([#59](https://github.com/spring-media/aws-lambda-router/pull/59)) (thanks to [@snorberhuis](https://github.com/snorberhuis))
-   build: updates build nodeJS version to 10

### 0.9.1 (September 11, 2020)

-   proxyIntegration: Allow defaultHeaders to be added to the OPTIONS request ([#53](https://github.com/spring-media/aws-lambda-router/pull/53)) (thanks again to [@TerryMooreII](https://github.com/TerryMooreII) and [LiveOakLabs](https://github.com/LiveOakLabs))

### 0.9.0 (September 01, 2020)

-   proxyIntegration: Adds customizable CORS configurations ([#52](https://github.com/spring-media/aws-lambda-router/pull/52)) (thanks to [@TerryMooreII](https://github.com/TerryMooreII) and [LiveOakLabs](https://github.com/LiveOakLabs))

### 0.8.4 (August 01, 2020)

-   proxyIntegration: expose route path ([#49](https://github.com/spring-media/aws-lambda-router/pull/49)) (thanks to [@evgenykireev](https://github.com/evgenykireev))

### 0.8.3 (May 07, 2020)

-   added records to the SQS ([#43](https://github.com/spring-media/aws-lambda-router/pull/43)) and SNS ([#44](https://github.com/spring-media/aws-lambda-router/pull/44)) action for further processing

### 0.8.2 (January 28, 2020)

-   added support for Open API parameter definitions e.g.: /section/{id}

### 0.8.1 (January 24, 2020)

-   fix: changed ProxyIntegrationEvent body type to be generic but defaults to unknown
-   fix: changed @types/aws-lambda from devDependency to dependency
-   **breaking**: error response objects (thrown or rejected) now need to set `statusCode` instead of `status` (consistent with response)

### 0.7.1 (January 22, 2020)

-   code style cleanup
-   fix: hosted package on npmjs should now worked

### 0.7.0 (January 17, 2020)

-   migrate to typescript
-   using @types/aws-lambda typings
-   proxyIntegration: cors is now optional (default: false)
-   removed use of aws lambda handler callback function (using Promise instead)
-   experimental _proxy path support_ (thanks to [@swaner](https://github.com/swaner))

### 0.6.2 (May 03, 2019)

-   take away old gulp dependency to run tests, works now with scripts in package.json
-   normalize request path to start from local host (thanks to [@napicella](https://github.com/napicella))

### 0.6.1 (January 02, 2019)

-   s3: fix: aggregate result promises to one promise;
-   s3: s3Route interface

### 0.6.0 (December 24, 2018)

-   new feature: S3 routes available.

### 0.5.0 (November 18, 2018)

-   new feature: SQS route integration now available;
-   bugfix: SNS integration now works with Array of message instead of single message

### 0.4.0 (September 13, 2018)

-   now [the Context Object](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-handler.html) pass through

### 0.3.1 (August 28, 2018)

-   proxyIntegration: avoid error if response object is not set
-   add some debug logging

### 0.3.0 (August 28, 2018)

-   proxyIntegration: add PATCH method;
-   allow for custom status codes from route (thanks to [@mintuz](https://github.com/mintuz))

### 0.2.3 (October 13, 2017)

-   better errorhandling (from PR #3)

### 0.2.2 (November 25, 2016)

-   proxyIntegration: set correct header values now for CORS

### 0.2.1 (November 25, 2016)

-   proxyIntegration: CORS in Preflight, status code 400 for invalid body, set more CORS headers as default

### 0.2.0 (November 21, 2016)

-   _Attention_: breaking changes for configuration;
-   add SNS event process

### 0.1.0 (November 15, 2016)

-   make it work now

### 0.0. (September 29, 2016)

-   initial release


lib/proxyIntegration.ts
import { APIGatewayEventRequestContext, APIGatewayProxyEventV2, APIGatewayProxyStructuredResultV2 } from 'aws-lambda';

import { HttpMethod, ProcessMethod } from './EventProcessor';
import { addCorsHeaders, CorsOptions } from './cors';

type ProxyIntegrationParams = {
    paths?: { [paramId: string]: string };
    routePath?: string;
};

type ProxyIntegrationBody<T = unknown> = {
    body: T;
};

type ErrorHandler = (
    error?: Error,
    request?: APIGatewayProxyEventV2,
    context?: APIGatewayEventRequestContext,
) => Promise<APIGatewayProxyStructuredResultV2 | void> | APIGatewayProxyStructuredResultV2 | void;
export type ProxyIntegrationEvent<T = unknown> = Omit<APIGatewayProxyEventV2, 'body'> &
    ProxyIntegrationParams &
    ProxyIntegrationBody<T>;
export type ProxyIntegrationResult = Omit<APIGatewayProxyStructuredResultV2, 'statusCode'> & {
    statusCode?: APIGatewayProxyStructuredResultV2['statusCode'];
};

export interface ProxyIntegrationRoute {
    path: string;
    method: HttpMethod;
    action: (
        request: ProxyIntegrationEvent<any>,
        context: APIGatewayEventRequestContext,
    ) => ProxyIntegrationResult | Promise<ProxyIntegrationResult> | string | Promise<string>;
}

export type ProxyIntegrationErrorMapping = {
    [reason: string]: APIGatewayProxyStructuredResultV2['statusCode'];
};

export type ProxyIntegrationError =
    | {
          statusCode: APIGatewayProxyStructuredResultV2['statusCode'];
          message: string;
      }
    | {
          reason: string;
          message: string;
      };

export interface ProxyIntegrationConfig {
    onError?: ErrorHandler;
    cors?: CorsOptions | boolean;
    routes: ProxyIntegrationRoute[];
    removeBasePath?: boolean;
    debug?: boolean;
    errorMapping?: ProxyIntegrationErrorMapping;
    defaultHeaders?: APIGatewayProxyStructuredResultV2['headers'];
    proxyPath?: string;
}

const NO_MATCHING_ACTION = (request: ProxyIntegrationEvent) => {
    throw {
        reason: 'NO_MATCHING_ACTION',
        message: `Could not find matching action for ${request.requestContext.http.path} and method ${request.requestContext.http.method}`,
    };
};

const processActionAndReturn = async (
    actionConfig: Pick<ProxyIntegrationRoute, 'action'>,
    event: ProxyIntegrationEvent,
    context: APIGatewayEventRequestContext,
    headers: APIGatewayProxyStructuredResultV2['headers'],
) => {
    const res = await actionConfig.action(event, context);
    if (!res || typeof res !== 'object' || typeof res.body !== 'string') {
        return {
            statusCode: 200,
            headers,
            body: JSON.stringify(res) || '{}',
        };
    }

    return {
        statusCode: 200,
        ...res,
        headers: {
            ...headers,
            ...res.headers,
        },
    };
};

export const process: ProcessMethod<
    ProxyIntegrationConfig,
    APIGatewayProxyEventV2,
    APIGatewayEventRequestContext,
    APIGatewayProxyStructuredResultV2
> = (proxyIntegrationConfig, event, context) => {
    if (proxyIntegrationConfig.debug) {
        console.log('Lambda proxyIntegrationConfig: ', proxyIntegrationConfig);
        console.log('Lambda event: ', event);
        console.log('Lambda context: ', context);
    }

    //validate config
    if (!Array.isArray(proxyIntegrationConfig.routes) || proxyIntegrationConfig.routes.length < 1) {
        throw new Error('proxyIntegration.routes must not be empty');
    }

    // detect if it's an http-call at all:
    if (!event.requestContext.http.method || !event.requestContext.http.path) {
        return null;
    }

    const headers: APIGatewayProxyStructuredResultV2['headers'] = proxyIntegrationConfig.cors
        ? addCorsHeaders(proxyIntegrationConfig.cors, event)
        : {};

    if (event.requestContext.http.method === 'OPTIONS') {
        Object.assign(headers ?? {}, proxyIntegrationConfig.defaultHeaders);
        return Promise.resolve({
            statusCode: 200,
            headers,
            body: '',
        });
    }

    Object.assign(headers ?? {}, { 'Content-Type': 'application/json' }, proxyIntegrationConfig.defaultHeaders);

    // assure necessary values have sane defaults:
    const errorMapping = proxyIntegrationConfig.errorMapping || {};
    errorMapping['NO_MATCHING_ACTION'] = 404;

    if (proxyIntegrationConfig.proxyPath) {
        event.requestContext.http.path = (event.pathParameters || {})[proxyIntegrationConfig.proxyPath];
        if (proxyIntegrationConfig.debug) {
            console.log(`proxy path is set: ${proxyIntegrationConfig.proxyPath}`);
            console.log(`proxy path with event path: ${event.requestContext.http.path}`);
        }
    } else {
        event.requestContext.http.path = normalizeRequestPath(event, proxyIntegrationConfig.removeBasePath);
    }

    try {
        const httpMethod = event.requestContext.http.method as HttpMethod;
        const actionConfig = findMatchingActionConfig(
            httpMethod,
            event.requestContext.http.path,
            proxyIntegrationConfig,
        ) || {
            action: NO_MATCHING_ACTION,
            routePath: undefined,
            paths: undefined,
        };

        const proxyEvent: ProxyIntegrationEvent = event as ProxyIntegrationEvent;

        proxyEvent.paths = actionConfig.paths;
        proxyEvent.routePath = actionConfig.routePath;
        if (event.body) {
            try {
                proxyEvent.body = JSON.parse(event.body);
            } catch (parseError) {
                console.log(`Could not parse body as json: ${event.body}`, parseError);
                return {
                    statusCode: 400,
                    headers,
                    body: JSON.stringify({
                        message: 'body is not a valid JSON',
                        error: 'ParseError',
                    }),
                };
            }
        }
        return processActionAndReturn(actionConfig, proxyEvent, context, headers).catch(async error => {
            console.log('Error while handling action function.', error);
            if (proxyIntegrationConfig.onError) {
                const result = await proxyIntegrationConfig.onError(error, event, context);
                if (result != undefined) {
                    return { headers, ...result };
                }
            }

            return convertError(error, errorMapping, headers);
        });
    } catch (error) {
        console.log('Error while evaluating matching action handler', error);

        if (proxyIntegrationConfig.onError) {
            const promise = proxyIntegrationConfig.onError(error as Error, event, context);
            Promise.resolve(promise).then(result => {
                if (result != undefined) {
                    return { headers, ...result };
                }

                return convertError(error as Error, errorMapping, headers);
            });
        }

        return convertError(error as Error, errorMapping, headers);
    }
};

const normalizeRequestPath = (event: APIGatewayProxyEventV2, removeBasePath: boolean = true): string => {
    if (isLocalExecution(event)) {
        return event.requestContext.http.path;
    }
    // ugly hack: if host is from API-Gateway 'Custom Domain Name Mapping', then event.path has the value '/basepath/resource-path/'
    // if host is from amazonaws.com, then event.path is just '/resource-path':
    const apiId = event.requestContext ? event.requestContext.apiId : null; // the apiId that is the first part of the amazonaws.com-host
    if (
        apiId &&
        event.headers &&
        event.headers.Host &&
        event.headers.Host.substring(0, apiId.length) !== apiId &&
        removeBasePath
    ) {
        // remove first path element:
        const groups = /\/[^\/]+(.*)/.exec(event.requestContext.http.path) || [null, null];
        return groups[1] || '/';
    }

    return event.requestContext.http.path;
};

const hasReason = (error: any): error is { reason: string } => typeof error.reason === 'string';
const hasStatus = (error: any): error is { statusCode: number } => typeof error.statusCode === 'number';

const convertError = (
    error: ProxyIntegrationError | Error,
    errorMapping?: ProxyIntegrationErrorMapping,
    headers?: APIGatewayProxyStructuredResultV2['headers'],
) => {
    if (hasReason(error) && errorMapping && errorMapping[error.reason]) {
        return {
            statusCode: errorMapping[error.reason],
            body: JSON.stringify({ message: error.message, error: error.reason }),
            headers,
        };
    } else if (hasStatus(error)) {
        return {
            statusCode: error.statusCode,
            body: JSON.stringify({ message: error.message, error: error.statusCode }),
            headers: addCorsHeaders({}, {} as APIGatewayProxyEventV2),
        };
    }
    try {
        return {
            statusCode: 500,
            body: JSON.stringify({
                error: 'ServerError',
                message: `Generic error:${JSON.stringify(error)}`,
            }),
            headers: addCorsHeaders({}, {} as APIGatewayProxyEventV2),
        };
    } catch (stringifyError) {}

    return {
        statusCode: 500,
        body: JSON.stringify({ error: 'ServerError', message: 'Generic error' }),
    };
};

const findMatchingActionConfig = (
    httpMethod: HttpMethod,
    httpPath: string,
    routeConfig: ProxyIntegrationConfig,
): (ProxyIntegrationRoute & ProxyIntegrationParams) | null => {
    const paths: ProxyIntegrationParams['paths'] = {};
    const matchingMethodRoutes = routeConfig.routes.filter(route => route.method === httpMethod);
    for (const route of matchingMethodRoutes) {
        if (routeConfig.debug) {
            console.log(`Examining route ${route.path} to match ${httpPath}`);
        }
        const pathPartNames = extractPathNames(route.path);
        const pathValues = extractPathValues(route.path, httpPath);
        if (pathValues && pathPartNames) {
            for (let ii = 0; ii < pathValues.length; ii++) {
                paths[pathPartNames[ii]] = decodeURIComponent(pathValues[ii]);
            }
            if (routeConfig.debug) {
                console.log(`Found matching route ${route.path} with paths`, paths);
            }
            return {
                ...route,
                routePath: route.path,
                paths,
            };
        }
    }
    if (routeConfig.debug) {
        console.log(`No match for ${httpPath}`);
    }

    return null;
};

const extractPathValues = (pathExpression: string, httpPath: string) => {
    const pathExpressionPattern = pathExpression.replace(/{[\w]+}|:[\w]+/g, '([^/]+)');
    const pathValueRegex = new RegExp(`^${pathExpressionPattern}$`);
    const pathValues = pathValueRegex.exec(httpPath);
    return pathValues && pathValues.length > 0 ? pathValues.slice(1) : null;
};

const extractPathNames = (pathExpression: string) => {
    const pathExpressionPattern = pathExpression.replace(/{[\w.]+}|:[\w.]+/g, '[:{]([\\w]+)}?');
    const pathNameRegex = new RegExp(`^${pathExpressionPattern}$`);
    const pathNames = pathNameRegex.exec(pathExpression);
    return pathNames && pathNames.length > 0 ? pathNames.slice(1) : null;
};

const isLocalExecution = (event: APIGatewayProxyEventV2) => {
    return (
        event.headers &&
        event.headers.Host &&
        (event.headers.Host.startsWith('localhost') || event.headers.Host.startsWith('127.0.0.1'))
    );
};
